src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/HKDF.cs

```diff
@@ -0,0 +1,255 @@
+            ReadOnlySpan<byte> saltSpan = salt ?? ReadOnlySpan<byte>.Empty;
+
+            byte[] prk = new byte[hashLength];
+
+            Extract(hashAlgorithmName, hashLength, ikm, saltSpan, prk);
+            return prk;
+        }
+
+        /// <summary>
+        /// Performs the HKDF-Extract function.
+        /// See section 2.2 of <a href="https://tools.ietf.org/html/rfc5869#section-2.2">RFC5869</a>
+        /// </summary>
+        /// <param name="hashAlgorithmName">Hash algorithm used for HMAC operations</param>
+        /// <param name="ikm">Input keying material</param>
+        /// <param name="salt">Salt value (a non-secret random value)</param>
+        /// <param name="prk">Buffer representing output pseudo-random key (prk)</param>
+        /// <returns>Number of bytes written to <paramref name="prk"/> buffer.</returns>
+        public static int Extract(HashAlgorithmName hashAlgorithmName, ReadOnlySpan<byte> ikm, ReadOnlySpan<byte> salt, Span<byte> prk)
+        {
+            int hashLength = HashLength(hashAlgorithmName);
+
+            if (prk.Length > hashLength)
+            {
+                prk = prk.Slice(0, hashLength);
+
+                // if it's too short we will throw later
+            }
+
+            Extract(hashAlgorithmName, hashLength, ikm, salt, prk);
+            return hashLength;
+        }
+
+        private static void Extract(HashAlgorithmName hashAlgorithmName, int hashLength, ReadOnlySpan<byte> ikm, ReadOnlySpan<byte> salt, Span<byte> prk)
+        {
+            if (prk.Length != hashLength)
+            {
+                throw new ArgumentException(nameof(prk));
+            }
+
+            Debug.Assert(HashLength(hashAlgorithmName) == hashLength);
+
+            using (IncrementalHash hmac = IncrementalHash.CreateHMAC(hashAlgorithmName, salt))
+            {
+                hmac.AppendData(ikm);
+                GetHashAndReset(hmac, prk);
+            }
+        }
+
+        /// <summary>
+        /// Performs the HKDF-Expand function
+        /// See section 2.3 of <a href="https://tools.ietf.org/html/rfc5869#section-2.3">RFC5869</a>
+        /// </summary>
+        /// <param name="hashAlgorithmName">Hash algorithm used for HMAC operations</param>
+        /// <param name="prk">Pseudorandom key of at least <see cref="HashLength"/> bytes (usually the output from Expand step)</param>
+        /// <param name="outputLength">Length of the output keying material</param>
+        /// <param name="info">Optional context and application specific information</param>
+        /// <returns>Output keying material</returns>
+        public static byte[] Expand(HashAlgorithmName hashAlgorithmName, byte[] prk, int outputLength, byte[] info = null)
+        {
+            if (prk == null)
+                throw new ArgumentNullException(nameof(prk));
+
+            int hashLength = HashLength(hashAlgorithmName);
+
+            if (outputLength <= 0 || outputLength > 255 * hashLength)
```

Comment:
The constant here should have a comment explaining that it comes from section 2.3 (the constraint on L in the Inputs section).

The upper bounds check should be done by division to avoid an overflow problem.

Reply:
can you clarify the overflow issue? the right hand side is constant for a given hash length

Reply:
> can you clarify the overflow issue?

Yep, there is no overflow issue (assuming we never have a hash algorithm that writes an 8MB output... which is a pretty good assumption).

Whoops.